### ISSUE TITLE
request: use mixins instead of inheritance

### ISSUE THREAD

**Issue by moonmaster9000** [zendeskarchive/broadcast#3]
I'd love to be able to turn in any existing class into a message by mixing in a module. In other words, instead of:

```
class Poke < Broadcast::Message
  medium :jabber

  def body
    "Poke!"
  end
end
```

I want to be able to do:

```
class Poke 
  include Broadcast::Message
  medium :jabber

  def body
    "Poke!"
  end
end
```

**Comment by marcinbunsch** [comment]
Hi Matt

Great idea, I did a quick experiment to see if it can be done without breaking the existing API (so Broadcast::Message is still there, but it includes Broadcast::Publishable so your class can to the same):
https://github.com/futuresimple/broadcast/commit/fbc5b3b6fdbc965e9dd93b97a3bbac6d842e978e

With this your example would look like this:

``` ruby
class Poke 
  include Broadcast::Publishable
  medium :jabber

  def body
    "Poke!"
  end
end
```

The only bit I'm not sure is that there's a constructor in the module, which I don't like. Suppose your class has it's own constructor, it might be problematic. I think the solution would be to not to have the constructor and leave it to your class to implement one should you require the ability to pass options to it. Would that be the behavior you would expect?

**Comment by moonmaster9000** [comment]
Hi Marcin, 

Thanks for the quick response! Yeah, the module really can't contain an initializer. And as for the the options, there should be several ways to set them IMO: 
1. At the class level, so that it's applied to all broadcasts for that class.
2. At the instance level, so that it's applied to all broadcasts for that particular instance
3. At the method level, so that the options can be set for a single publish

The first might look like this:

```
class Message
  include Broadcast::Publishable
  broadcast_options :blah => :blah
  #....
end
```

The second might look like this:

```
class Message
  include Broadcast::Publishable
  #...
end

m = Message.new
m.broadcast_options = {:blah => :blah}
```

The third might look like this:

```
class Message
  include Broadcast::Publishable
  #....
end

m = Message.new
m.publish :blah => :blah
```

**Comment by marcinbunsch** [comment]
These are really cool ideas, but I kind of feel that this adds extra overhead that just is not required when you include the module. The thing is that when would you want to use the mixin - usually when you have a class that's doing something already and you want to publish what it is doing. Consider the following example:

``` ruby

class User < ActiveRecord::Base
  include Broadcast::Publishable

  after_create :publish

  medium :irc
  medium :jabber
  medium :yammer

  def body
    "#{name} just signed up!"
  end

end

User.create(:name => "John") # => publishes to irc, jabber and yammer

```

So we have a User model which is also a message, which gets sent on save. Introducing an extra accessor is unnecessary as we take all data from the model's attributes. 

Essentially my thinking is that this that when you use a mixin, you use it on some existing class or you want to mix something else, so you already have a source of data to put in the message.

I hope I didn't mess up too much :) Would that solution be correct for your use case?

**Comment by moonmaster9000** [comment]
hi, sorry, i really don't have a usecase yet; I think i was confused though by the "options" created on instantiation. Can you provide examples of what options you might provide on instantiation, and how those would be applied to the broadcast?

**Comment by marcinbunsch** [comment]
I think you're right that the name 'options' is confusing. In Broadcast::Message instances, 'options' is basically a hash with some values sent during instantation so that you don't have to . Let's consider the following example:

``` ruby
class Signup < Brodcast::Message
  medium :irc

  def body
    user = User.find(options.user_id)
    "User #{user.name} has signed up"
  end
end

# Run it later like this
Signup.new(:user_id => 4).publish
```

So the goal of having this hash is to allow you to create more reusable classes inheriting from Broadcast::Message. But my reasoning is that when you include the module, you have an idea what you want from the class, so there is no need to enforce such storage for 'configuration' as this 'configuration' should come from the class, so it's up to the end developer to decide what methods to call in the 'body' method.

Do check out the 'mixins' branch, it has that implementation, it has that implementation without the 'options' in the module.

**Comment by moonmaster9000** [comment]
oh I see, I had assumed the options were related to options for the broadcast (in my mind, I imagined options like :retry => true and :on_fail => :method)

**Comment by marcinbunsch** [comment]
Yeah, that's why I'm kicking myself for picking that name :)

**Comment by moonmaster9000** [comment]
hehe no worries, i've been there. that's what major version bumps are for! :-)

would you ever see adding broadcast options (like :retry and :on_fail) that are definable on a per class, per instance, per publish basis?

**Comment by marcinbunsch** [comment]
Hi

Just wanted to let you know that I merged in the mixins branch into broadcast 0.3.0, hope you find it useful! :)
